scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+
+cdef validate_qr(object q0, object r0, bint overwrite_q, int q_order,
+                 bint overwrite_r, int r_order, bint chkfinite):
+    cdef cnp.ndarray Q
+    cdef cnp.ndarray R
+    cdef int typecode
+    cdef bint economic = False
+
+    q_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES  
+    r_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES
+
+    if not overwrite_q:
+        q_order |= cnp.NPY_ENSURECOPY
+
+    if not overwrite_r:
+        r_order |= cnp.NPY_ENSURECOPY
+
+    # in the interests of giving better error messages take any number of
+    # dimensions here.
+    Q = PyArray_CheckFromAny(q0, NULL, 0, 0, q_order, NULL)
+    R = PyArray_CheckFromAny(r0, NULL, 0, 0, r_order, NULL)
+
+    if Q.ndim != 2 or R.ndim != 2:
+        raise ValueError('Q and R must be 2d')
+
+    typecode = cnp.PyArray_TYPE(Q)
+
+    if typecode != cnp.PyArray_TYPE(R):
+        raise ValueError('q and r must have the same type')
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE 
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('only floatingcomplex arrays supported')
+
+    # we support MxM MxN and MxN NxN
+    if Q.shape[1] != R.shape[0]:
+        raise ValueError('Q and R do not have compatible shapes')
+
+    # so one or the other or both should be square.
+    if Q.shape[0] != Q.shape[1] and R.shape[0] == R.shape[1]:
+        economic = True
+    elif Q.shape[0] != Q.shape[1]:
+        raise ValueError('bad shapes.')
+
+    Q = validate_array(Q, chkfinite)
+    R = validate_array(R, chkfinite)
+
+    return Q, R, typecode, Q.shape[0], R.shape[1], economic
+ 
+cdef void* extract(cnp.ndarray a, int* as):
+    if a.ndim == 2:
+        as[0] = a.strides[0] / cnp.PyArray_ITEMSIZE(a)
+        as[1] = a.strides[1] / cnp.PyArray_ITEMSIZE(a)
+    elif a.ndim == 1:
+        as[0] = a.strides[0] / cnp.PyArray_ITEMSIZE(a)
+        as[1] = 0
+    return cnp.PyArray_DATA(a)
+
+@cython.embedsignature(True)
+def qr_delete(Q, R, k, p=1, which='row', overwrite_qr=True, check_finite=True):
+    """QR downdate on row or column deletions
+
+    If ``A = Q R`` is the qr factorization of A, return the qr factorization
+    of `A` where `p` rows or columns have been removed starting at row or
+    column `k`.
+
+    Parameters
+    ----------
+    Q : (M, M) or (M, N) array_like
+        Unitary/orthogonal matrix from QR decomposition.
+    R : (M, N) or (N, N) array_like
+        Upper trianglar matrix from QR decomposition.
+    k : int
+        index of the first row or column to delete.
+    p : int, optional
+        number of rows or columns to delete, defaults to 1.
+    which: {'row', 'col'}, optional
+        Determines if rows or columns will be deleted, defaults to 'row'
+    overwrite_qr : bool, optional
+        If True, consume Q and R, overwriting their contents with their
+        downdated versions, and returning approriately sized views.  
+        Defaults to True.
```

Comment:
``qr`` and ``qr_multiply`` default to False. Both for consistency and safety that should probably be the default here as well.